### PR TITLE
loops-simple_programs_with_repetition-hard-q4

### PR DESCRIPTION
**Pull request by SamamthaH**
…ument question code. Included old code in a comment, but waiting to see if the question will be moved to a different section.

## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/loops/simple_programs_with_repitition/hard/q4/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class MainTest extends BaseTest {
+    //Java
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q4/MainTest.java:18]:
```suggestion
    // Java
```

File: src/loops/simple_programs_with_repitition/hard/q4/Reduceros.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package loops.simple_programs_with_repitition.hard.q4;
+
+/*
+Write a program to display a "single-digit reduction" of a number taken from the user by adding up its digits until
+a single-digit number is obtained. Eg. 567 becomes 5 + 6 + 7 = 18 = 1 + 8 => 9)
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q4/Reduceros.java:5]:
```suggestion
a single-digit number is obtained. Eg. 567 becomes 5 + 6 + 7 = 18 = 1 + 8 => 9). Ensure your output matches the sample. 
```

File: src/loops/simple_programs_with_repitition/hard/q4/Reduceros.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            n = 0;
+            for (int x = 0; x < s.length(); x++) {
+                n += Integer.parseInt(String.valueOf(s.charAt(x)));
+            }
+        }
+        System.out.println(nn + "'s single number reduction is " + n);
+    }
+}
+
+/*
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q4/Reduceros.java:32]: I think we should have the original solution code be the one this question runs off of.

File: src/loops/simple_programs_with_repitition/hard/q4/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package loops.simple_programs_with_repitition.hard.q4;
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q4/MainTest.java:1]: Move package

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/simple_programs_with_repetition/hard/q4/Reduceros.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+/*
+Write a program to display a "single-digit reduction" of a number taken from the user by adding up its digits until
+a single-digit number is obtained. Eg. 567 becomes 5 + 6 + 7 = 18 = 1 + 8 => 9). Ensure your output matches the sample.
+
+Sample Output:
+Enter a number:
+567
+567's single number reduction is 9
+ */
+
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repetition/hard/q4/Reduceros.java:12]: You should write a comment for this question that includes the reasoning behind the inclusion of this code. I know we talked about it in the pull request but if someone randomly stumbles upon this, I want it to make sense.

File: src/loops/simple_programs_with_repetition/hard/q4/Reduceros.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
+package loops.simple_programs_with_repetition.hard.q4;
+
+/*
+Write a program to display a "single-digit reduction" of a number taken from the user by adding up its digits until
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repetition/hard/q4/Reduceros.java:4]: We need to incorporate what to do with negative numbers in this question. If the question works with negative numbers, we should specify to include them and test them. If not, or if you think it doesn't make sense, we should add a test for them and they can be invalid input.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/simple_programs_with_repetition/hard/q4/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    public void runMain(){ Reduceros.main(new String[0]); }
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of("576", "9"), Arguments.of("0", "0"), Arguments.of("49723", "7"), Arguments.of("55", "1"), Arguments.of("-1", "invalid!"), Arguments.of("-10", "invalid!"), Arguments.of("-6785", "invalid!"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void calculatedReductionCorrectly(String in, String reduced) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repetition/hard/q4/MainTest.java:41]:
```suggestion
    void calculatesReductionCorrectly(String in, String reduced) throws InvalidClauseException {
```

File: src/loops/simple_programs_with_repetition/hard/q4/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        TestOption.incorrectStructureErrorMessage = "Your program does not calculate the reduction properly.";
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repetition/hard/q4/MainTest.java:42]:
```suggestion
        TestOption.incorrectStructureErrorMessage = "Your program does not calculate the single-digit reduction properly.";
```

File: src/loops/simple_programs_with_repetition/hard/q4/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        runWithInput(in, new Clause[]{
+                new StringLiteral(reduced)
+        });
+        assertEquals(getItemByName("number"), in, "Your program does not output the same number that was input as per the sample output.");
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repetition/hard/q4/MainTest.java:46]:
```suggestion
        assertEquals(getItemByName("number"), in, "Your program does not print the correct output.");
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]
